### PR TITLE
wgengine/monitor: make Close not block forever on Linux

### PR DESCRIPTION
**Pull request by bradfitz**

### CODE REVIEW

File: wgengine/monitor/monitor_linux.go [hunk 1]
```diff
@@ -45,6 +46,7 @@ func newOSMon() (osMon, error) {
 }
 
 func (c *nlConn) Close() error {
+	c.conn.SetDeadline(time.Unix(0, 0)) // abort any Receive in flight
```
Comment (danderson) [inline wgengine/monitor/monitor_linux.go]: Is this the idiom for a zero time? I've always used `time.Time{}`.

**Comment by bradfitz** [comment]
This fixes the deadlock that was always there but the recent changes tested (and were flaky as a result).

File: wgengine/monitor/monitor_linux.go (hunk 1 above)
Comment (danderson) [reply wgengine/monitor/monitor_linux.go]: Also confused: I thought that Close() unblocked all readers. Clearly not? Is this a new behavior?

Comment (bradfitz) [reply wgengine/monitor/monitor_linux.go]: > Is this the idiom for a zero time? I've always used time.Time{}.

It's not a zero time. It's just some time that's in the past, and with the added benefit that it's a time that's safe to convert into a C time_t if needed, if it happens to go that route (and I was too lazy to check).

> Also confused: I thought that Close() unblocked all readers. Clearly not? Is this a new behavior?

The code comments in the package suggested it did (that that was the intention) but it ultimately called through to a sysSocket.Close implementation a few layers down that did have a mutex and deadlocked. There's an open bug (https://github.com/mdlayher/netlink/issues/162) too about this.

**Pull request merged**